### ISSUE TITLE
C++ exceptions cause 'segmentation fault: 11' in PHP wrapper.

### ISSUE THREAD

**Issue by SergeySeroshtan** [swig/swig#240]
Reproduced on SWIG version 3.0.2.
###### Steps to reproduce:
 1. Write value-object C++ class with copy constructor.
 1. Write class with method that conforms to conditions described below.
 1. Write SWIG interface file with standard exception handling code.
 1. Write PHP code that invokes method which throws exception.
 1. Run PHP code.

###### Wrapped method conditions:
- wrapped method should return object by value;
- returned object should contain copy constructor;
- wrapped method should throw exception before resulting object will be returned.

###### Reason:
Default SWIG macro function 'SWIG_exception' defined for PHP as:
```
#define SWIG_exception(code, msg) zend_throw_exception(NULL, (char*)msg, code TSRMLS_CC)
```
**does not interrupt method's execution flow**, which leads to **dereferencing of NULL object** that returned as result.

###### Solution:
Redefine SWIG macro function 'SWIG_exception'  for PHP as:
```
#ifdef SWIGPHP
%{
#undef SWIG_exception
#define SWIG_exception(code, msg) { zend_throw_exception(NULL, (char*)msg, code TSRMLS_CC); return; }
%}
#endif
```

###### Example: modification of the SWIG example located at 'Examples/php/class'
example.h
```
/* File : example.h */

class Value {
public:
  Value(int a, int b);
  Value(const Value& other);
private:
  int a_;
  int b_;
};

class Shape {
public:
  Shape() {
    nshapes++;
  }
  virtual ~Shape() {
    nshapes--;
  }
  double  x, y;
  void    move(double dx, double dy);
  virtual double area() = 0;
  virtual double perimeter() = 0;
  virtual Value throwException();
  static  int nshapes;
};

class Circle : public Shape {
private:
  double radius;
public:
  Circle(double r) : radius(r) { }
  virtual double area();
  virtual double perimeter();
};

class Square : public Shape {
private:
  double width;
public:
  Square(double w) : width(w) { }
  virtual double area();
  virtual double perimeter();
};
```
example.cxx
```
/* File : example.cxx */

#include "example.h"
#define M_PI 3.14159265358979323846

#include <stdexcept>

Value::Value(int a, int b) : a_(a), b_(b) {}
Value::Value(const Value& other) : a_(other.a_), b_(other.b_) {}

/* Move the shape to a new location */
void Shape::move(double dx, double dy) {
  x += dx;
  y += dy;
}

Value Shape::throwException() {
    throw std::logic_error("Some custom error.");
    return Value(5, 10);
}

int Shape::nshapes = 0;

double Circle::area() {
  return M_PI*radius*radius;
}

double Circle::perimeter() {
  return 2*M_PI*radius;
}

double Square::area() {
  return width*width;
}

double Square::perimeter() {
  return 4*width;
}
```
example.i
```
/* File : example.i */
%module example

%{
#include "example.h"
%}

%include exception.i

/* Let's just grab the original header file here */
%exception {
    try {
        $action
    } catch (std::exception& e) {
      SWIG_exception(SWIG_SystemError, e.what());
    }
}

%include "example.h"
```

runme.php
```
<?php

# This example illustrates how member variables are wrapped.

require("example.php");

# ----- Object creation -----

print "Creating some objects:\n";
$c = new Circle(10);
print "    Created circle\n";
$s = new Square(10);
print "    Created square\n";

# ----- Access a static member -----

print "\nA total of " . Shape::nshapes() . " shapes were created\n";

# ----- Member data access -----

# Set the location of the object.
# Note: methods in the base class Shape are used since
# x and y are defined there.

$c->x = 20;
$c->y = 30;
$s->x = -10;
$s->y = 5;

print "\nHere is their current position:\n";
print "    Circle = ({$c->x},{$c->y})\n";
print "    Square = ({$s->x},{$s->y})\n";

# ----- Call some methods -----

# Notice how the Shape_area() and Shape_perimeter() functions really
# invoke the appropriate virtual method on each object.
print "\nHere are some properties of the shapes:\n";
foreach (array($c,$s) as $o) {
      print "    ". get_class($o) . "\n";
      print "        area      = {$o->area()}\n";
      print "        perimeter = {$o->perimeter()}\n";
}

# ----- Throw exception -----
try {
    $c->throwException();
} catch (Exception $e) {
    print ("Expected exception was thrown with message: " . $e->getMessage() . "\n");
}

# ----- Delete everything -----

print "\nGuess I'll clean up now\n";

# Note: this invokes the virtual destructor
$c = NULL;
$s = NULL;

# and don't forget the $o from the for loop above.  It still refers to
# the square.
$o = NULL;

print Shape::nshapes() . " shapes remain\n";
print "Goodbye\n";

?>
```

**Comment by ojwb** [comment]
@breznak This issue is specific to SWIG's PHP backend, so your problem is unrelated.  Please file a new issue with a self-contained example which reproduces the problem.

**Comment by ojwb** [comment]
Looks like this stalled due to lack of feedback from the submitter.

I've rebased my php-exception-fix branch onto current master, force-pushed it, and opened PR#785 so it gets CI tested.

As noted above, turning your demonstrator above into a new testcase under `Examples/php/test-suite/` would still be very useful.

**Comment by ojwb** [comment]
Fixed by 26bbc96d58b81293a2357ba3ba7c4419d87417dd.

**Comment by SergeySeroshtan** [comment]
Thanks a lot for this fix. I was very busy to provide PR with appropriate regression tests.
